### PR TITLE
Add ranked date and submitted date filtering to song select

### PR DESCRIPTION
**Pull request by WitherFlower**
Implements `ranked` and `submitted` filters, following the same logic as the osu-web search, which is described in https://github.com/ppy/osu-web/pull/7358

As a brief summary :
- Dates in the formats `yyyy`, `yyyy-mm` and `yyyy-mm-dd` are accepted.
- The valid separators are `-`, `/` and `.`
- Dates are treated as UTC to keep search results consistent between web and client.
- I did not add support for hours, minutes and seconds as I don't think it's really useful, but I can make the addition if deemed necessary

This is especially useful now that #29862 has been merged, as all players will have correct ranked/submitted date values in their databases starting from next release.

> P.S. : This is my first PR here, so please tell me if I horribly messed up some step of the process.

### CODE REVIEW

File: osu.Game/Screens/Select/Carousel/CarouselBeatmap.cs [hunk 1]
```diff
@@ -66,6 +66,8 @@ private bool checkMatch(FilterCriteria criteria)
             match &= !criteria.OverallDifficulty.HasFilter || criteria.OverallDifficulty.IsInRange(BeatmapInfo.Difficulty.OverallDifficulty);
             match &= !criteria.Length.HasFilter || criteria.Length.IsInRange(BeatmapInfo.Length);
             match &= !criteria.LastPlayed.HasFilter || criteria.LastPlayed.IsInRange(BeatmapInfo.LastPlayed ?? DateTimeOffset.MinValue);
+            match &= !criteria.DateRanked.HasFilter || criteria.DateRanked.IsInRange(BeatmapInfo.BeatmapSet?.DateRanked ?? DateTimeOffset.MinValue);
+            match &= !criteria.DateSubmitted.HasFilter || criteria.DateSubmitted.IsInRange(BeatmapInfo.BeatmapSet?.DateSubmitted ?? DateTimeOffset.MinValue);
```
[the remark this answers is not in the document]
Comment (WitherFlower) [reply osu.Game/Screens/Select/Carousel/CarouselBeatmap.cs:70]: Date Ranked and Date Submitted sorts don't hide Unranked/Unsubmitted maps on purpose (see https://github.com/ppy/osu/pull/22775), so I didn't know if the filters should match this.

Personally I'd be for filtering out unranked maps, as having to add `status=r` in my query whenever I do `ranked<2010` to remove unranked maps feels redundant.

[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Screens/Select/Carousel/CarouselBeatmap.cs:70]: Just reading through #22775, it is certainly unrelated here as that PR fixes an issue where unranked/unsubmitted beatmaps are being filtered out when using a specific *sorting mode*. In this case, we're filtering out beatmaps, so it doesn't make sense to not filter out unranked/unsubmitted beatmaps when specifying a specific range of ranked/submitted beatmaps.

I agree with the suggestion above.

[the remark this answers is not in the document]
Comment (WitherFlower) [reply osu.Game/Screens/Select/Carousel/CarouselBeatmap.cs]: Applied the proposed change. Feels better to use in-game indeed.

**Review by bdach — APPROVED** [review]

**Review by peppy — APPROVED** [review]
Seems good 👍

**Pull request merged**